packages/datetime/src/dateRangeInput.tsx

```diff
@@ -256,14 +276,219 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
     private handleDateRangePickerChange = (selectedRange: DateRange) => {
         if (this.props.value === undefined) {
             const [selectedStart, selectedEnd] = fromDateRangeToMomentDateRange(selectedRange);
-            this.setState({ selectedStart, selectedEnd });
+
+            let isStartInputFocused: boolean;
+            let isEndInputFocused: boolean;
+
+            let startHoverString: string;
+            let endHoverString: string;
+
+            if (isMomentNull(selectedStart)) {
+                // focus the start field by default or if only an end date is specified
+                isStartInputFocused = true;
+                isEndInputFocused = false;
+
+                // for clarity, hide the hover string until the mouse moves over a different date
+                startHoverString = null;
+            } else if (isMomentNull(selectedEnd)) {
+                // focus the end field if a start date is specified
+                isStartInputFocused = false;
+                isEndInputFocused = true;
+
+                endHoverString = null;
+            } else if (this.state.mostRecentlyFocusedField === DateRangeBoundary.START) {
+                // keep the start field focused
+                isStartInputFocused = true;
+                isEndInputFocused = false;
+            } else {
+                // keep the end field focused
+                isStartInputFocused = false;
+                isEndInputFocused = true;
+            }
+
+            this.setState({
+                selectedEnd,
+                selectedStart,
+                isEndInputFocused,
+                isStartInputFocused,
+                startHoverString,
+                endHoverString,
+                endInputString: this.getFormattedDateString(selectedEnd),
+                startInputString: this.getFormattedDateString(selectedStart),
+                wasLastFocusChangeDueToHover: false,
+            });
         }
         Utils.safeInvoke(this.props.onChange, selectedRange);
     }
 
+    private handleDateRangePickerHoverChange = (hoveredRange: DateRange) => {
+        if (hoveredRange == null) {
+            // undo whatever focus changes we made while hovering
+            // over various calendar dates
+            const isEndInputFocused = (this.state.preferredBoundaryToModify === DateRangeBoundary.END);
+            const isStartInputFocused = !isEndInputFocused;
+            const mostRecentlyFocusedField = (isEndInputFocused) ? DateRangeBoundary.END : DateRangeBoundary.START;
```

Comment:
I haaaaate that I had to set `mostRecentlyFocusedField` here. The behavior works fine to the human eye without this, but some tests were failing because `handleInputFocus` wasn't being invoked before the test assertion ran. Thus, the `is{Start,End}InputFocused` variables weren't being updated in time, so downstream code that referred to these variables was doing incorrect intermediate things before ultimately correcting itself. Ugh.